libraries/shared/logs/extractor.go

```diff
@@ -0,0 +1,156 @@
+	}
+
+	if extractor.StartingBlock == nil {
+		extractor.StartingBlock = &config.StartingBlockNumber
+	} else if earlierStartingBlockNumber(config.StartingBlockNumber, *extractor.StartingBlock) {
+		extractor.StartingBlock = &config.StartingBlockNumber
+	}
+
+	addresses := transformer.HexStringsToAddresses(config.ContractAddresses)
+	extractor.Addresses = append(extractor.Addresses, addresses...)
+	extractor.Topics = append(extractor.Topics, common.HexToHash(config.Topic))
+	return nil
+}
+
+// Fetch and persist watched logs
+func (extractor LogExtractor) ExtractLogs(recheckHeaders constants.TransformerExecution) (error, bool) {
+	if len(extractor.Addresses) < 1 {
+		logrus.Errorf("error extracting logs: %s", ErrNoWatchedAddresses.Error())
+		return ErrNoWatchedAddresses, noMissingHeadersFound
+	}
+
+	uncheckedHeaders, uncheckedHeadersErr := extractor.CheckedHeadersRepository.UncheckedHeaders(*extractor.StartingBlock, -1, getCheckCount(recheckHeaders))
+	if uncheckedHeadersErr != nil {
+		logrus.Errorf("error fetching missing headers: %s", uncheckedHeadersErr)
+		return uncheckedHeadersErr, noMissingHeadersFound
+	}
+
+	if len(uncheckedHeaders) < 1 {
+		return nil, noMissingHeadersFound
+	}
+
+	for _, header := range uncheckedHeaders {
+		logs, fetchLogsErr := extractor.Fetcher.FetchLogs(extractor.Addresses, extractor.Topics, header)
+		if fetchLogsErr != nil {
+			logError("error fetching logs for header: %s", fetchLogsErr, header)
+			return fetchLogsErr, missingHeadersFound
+		}
+
+		if len(logs) > 0 {
+			transactionsSyncErr := extractor.Syncer.SyncTransactions(header.Id, logs)
+			if transactionsSyncErr != nil {
+				logError("error syncing transactions: %s", transactionsSyncErr, header)
+				return transactionsSyncErr, missingHeadersFound
+			}
+
+			createLogsErr := extractor.LogRepository.CreateHeaderSyncLogs(header.Id, logs)
+			if createLogsErr != nil {
+				logError("error persisting logs: %s", createLogsErr, header)
+				return createLogsErr, missingHeadersFound
+			}
+		}
+
+		markHeaderCheckedErr := extractor.CheckedHeadersRepository.MarkHeaderChecked(header.Id)
+		if markHeaderCheckedErr != nil {
+			logError("error marking header checked: %s", markHeaderCheckedErr, header)
+			return markHeaderCheckedErr, missingHeadersFound
+		}
+	}
+	return nil, missingHeadersFound
+}
+
+func earlierStartingBlockNumber(transformerBlock, watcherBlock int64) bool {
+	return transformerBlock < watcherBlock
+}
+
+func logError(description string, err error, header core.Header) {
+	logrus.WithFields(logrus.Fields{
+		"headerId":    header.Id,
+		"headerHash":  header.Hash,
+		"blockNumber": header.BlockNumber,
+	}).Errorf(description, err.Error())
+}
+
+func getCheckCount(recheckHeaders constants.TransformerExecution) int64 {
+	if recheckHeaders == constants.HeaderMissing {
+		return 1
+	} else {
+		return constants.RecheckHeaderCap
+	}
+}
+
+func (extractor *LogExtractor) updateCheckedHeaders(config transformer.EventTransformerConfig) error {
```

Comment:
Yeah this is definitely a little complicated - would welcome ideas about ways to simplify. The basic idea is that we can add new transformers after the system has already been running for awhile and it will automatically recheck previously-checked headers that are relevant to the new transformer.

Your understanding of the function is correct, and my thinking with marking logs checked here was that (1) it only happens after we've marked headers unchecked, so there's no risk of the log's being marked as checked preventing us from syncing those logs, and (2) marking logs as checked requires iterating through that address+topic0 combinations, which we're already doing in `AddTransformers` but not `ExtractLogs`.

Definitely happy to refactor/modify the approach - this was just my first pass at enabling new transformers to sync against already-checked headers with minimal additional code. What do you think?

Reply:
I think I understand - so the main motivation for marking logs as checked here in `updateCheckedHeaders` is for a performance gain in not having to iterate through the address + topic0 combos? That makes sense to me.  

I think I was mainly confused about marking logs as checked before they were unchecked, but maybe a comment explaining why we're doing this could help? Also what do you think about renaming `updateCheckedHeaders` to `updateCheckedHeadersAndLogs` or something?

Reply:
Also, marking marking both `headers` and `logs` as **checked**, and that we have a `header_sync_log` table and a `checked_logs` table tripped me up a bit as well. I don't know if I have any ideas on how this may be simplified, but just thought I'd throw it out there as a potential hang up for folks.

Reply:
Thinking I may change `checked_logs` to `watched_logs` or something similar - really the purpose of the table is just to keep track of what logs we were fetching when we marked a header as checked for logs